### ISSUE TITLE
body input parameter does not work with plain data types

### ISSUE THREAD

**Issue by ryo-nbx** [lukeautry/tsoa#95]
I found an issue with using the `Body` input parameter with non object data types. I tried to set some simple data changes (number, string values ...) over put API functions. Although the routes generation was successful the API methods where not called a single time when the URL of the functions was called (over the swagger UI). It seems that the generated code of the API and validation could not handle this simple data types. 

Here is an example that give some insight of what I found.

PersonController.ts
`@Put('{id}/name')`
`public async setPersonName (@Path() id: string, @Body() name: string ) : Promise<void> {`

Generation: successful

When calling the method over Swagger UI. name = ...

- `abc` : exception = `SyntaxError: Unexpected token a`
- `'abc'` : exception = `SyntaxError: Unexpected token '`
- `"abc"` : exception = `SyntaxError: Unexpected token "`
- `['abc']` : exception = `SyntaxError: Unexpected token ' in JSON at position 1`
- `["abc"]` : console output = `[object Object]` , no function call
- `{"abc"}` : exception = `SyntaxError: Unexpected token } in JSON at position 6`
- `{"value":"abc"}` : console output = `[object Object]` , no function call

The same happens when using `number` or `boolean` parameters. There is a small difference with number inputs:

- `[12]` : Is working!

You can't enter any working value in case of boolean inputs over the Swagger UI because it offers a drop down list (!) with values that will cause parse exceptions.

There is a simple workaround. You can define wrapper classes for those inputs and use them. For example for text input:

BodyInputs.ts
`export interface TextInput {`
`    "value": string;`
`}`

PersonController.ts
`@Put('{id}/name')`
`public async setPersonName (@Path() id: string, @Body() name: TextInput ) : Promise<void> {`

After route generation just put a simple JSON input object: `{"value": "abc"}`

There are issues with arrays as `Body` parameters too. 

Will there be a solution for this issue? If not, someone should document that it this kind of definition is not working for TSOA and provide a workaround sample for other users.

**Comment by ryo-nbx** [comment]
To detail the issue with arrays in `Body` parameters some details.

Swagger codegen generates controller functions like this:

PersonController.ts
`@Put('{id}/groups')`
`public async setGroupIds (@Path() id: string, @Body() groups: Array<number>) : Promise<void> {`

This code causes generation exception:
`There was a problem resolving type of 'Arraynumber'.`
`Error: Error generate parameter method: 'PersonController.setGroupIds' argument: groups Error: Only string indexers are supported`

To fix the issue one can change it to:

PersonController.ts
`@Put('{id}/groups')`
`public async setGroupIds (@Path() id: string, @Body() groups: number[]) : Promise<void> {`

Route generation is successful. Swagger UI tells you that the type is `Array[]`. But no input is accepted by the generated API. 

The previous workaround applies here too.

BodyInputs.ts
`export interface NumArrayInput {`
`    "value": number[];`
`}`

PersonController.ts
`@Put('{id}/groups')`
`public async setGroupIds (@Path() id: string, @Body() groups: NumArrayInput) : Promise<void> {`

working API call Body parameter: `{"value": [1,2,3,4,5,6,7,null,9,10]}`

**Comment by dgreene1** [comment]
I started to implement this and it suddenly occurred to me that the vast majority of tsoa users (myself included) use BodyParser which expects the body to be JSON. And since string and number are not valid  JSON, the following don't even come into the route in the way you would expect. So what I'm getting at is, should tsoa really support string and number as a body type?

Take for instance the string example you gave:
```ts
@Put('{id}/name')
public async setPersonName (@Path() id: string, @Body() name: string, @Request request: koa.Request ) : Promise<void> {
```
After passing through BodyParser, this is what would happen:
```ts
// let's say a user passed in "I am a string" for the body
//     inside of the route, you would find the following
console.log(request.body) // { "I am a string": undefined }
```
So in order for us to support `"string"` as a body type, we would:
* have to find a way to grab the rawBody off of the request (which is easy enough in Koa, but seems challenging in Express)
* we would have to expose a way for developers to specify the `'content-type"` since a string for the body is definitely not `"application/json"`. At best it would be`"application/x-www-form-urlencoded"`, `"multipart/form-data"`, `"text/plain; charset=utf-8"`, `"text/html"` but we can't know which one the developer intended.

In other words, this would be a decent amount of work to cover a case that is potentially under-utilized. So before I can take this up, can you or someone else clarify why you would want a string body to be sent in when there is an easy workaround?

**Workaround:**
Just send string in as a property on JSON. Here's an example using the setPersonName example you gave:
```ts
export interface IHaveAName {
  name: string
}

@Put('{id}/name')
public async setPersonName (@Path() id: string, @Body() body: IHaveAName) : Promise<IHaveAName> {
  console.log(body.name);
  return body;
}
```

**Comment by ryo-nbx** [comment]
I think you can close this topic. The main reason is the Open API specification. 

https://swagger.io/specification/#operationObject

The specification only supports objects as body parameter. No plain string or integer is allowed anymore. So there is no need to support plain datatypes as body parameters.

Though an important question should be answered. What version of OpenAPI should TSOA support / produce?

**Comment by ryo-nbx** [comment]
After looking further into the spec the matter seems more difficult than it seemed to me at the first look.

In both V2 and V3 OpenAPI spec the body can use the schema description. And by doing so you can use plain datatypes or strings or arrays as type. It looks to me that this data is not wrapped in an object by the spec. So the spec allows using plain datatypes as body parameters.

https://swagger.io/specification/v2/#schemaObject (v2)
https://swagger.io/specification/#requestBodyObject (v3)

The next consideration to think about is how the generated client code from swagger.io would handle body parameters with or without a workaround for this topic. Since this will be the counterpart that most people will use with TSOA generated APIs.

**Comment by dgreene1** [comment]
Yes @ryo-nbx. I’m aware of what you mentioned about the spec. The spec is wide and has many use cases. Not all of them are popular. 

The remaining problem is that I’m not aware of how to implement an Koa server that sometimes accepts JSON bodies and other times accepts string bodies. I described that problem in detail in comment https://github.com/lukeautry/tsoa/issues/95#issuecomment-533628700. But to summarize again, if you are using BodyParser you won’t be able to read any primitive bodies.

And since the tsoa automated integration tests use bodyparser, I can’t write automated tests that read primitive bodies. 

As a metaphor, a Koa server can not order chocolate and vanilla. It can only order one or the other.
